### PR TITLE
[#598] Expose native audio events

### LINKED ISSUE #598: Expose native Audio/Sound events through Excalibur
### Context

Currently, the audio/sound implementations in Excalibur do not expose the native dom events like "playing", "loaded", etc that may be useful when working with complicated sounds and workflows.

See here: http://kamranicus.com/wp81-ie11-audio-bug/ for an interesting example.
### Proposal

Once resources are refactored, `ex.Sound` should maybe implement the `EventDispatcher` interface and expose the native sound events that make sense.

``` typescript
var sound = new ex.Sound('mymp3.mp3');

// pass through the native event type, wrap the audio event in an ex type
sound.on('loaded', (e: ex.NativeSoundEvent) => { 
  // do work
});

sound.on('playing', (e: ex.NativeSoundEvent) => {
 // do work
});
```

### PR DESCRIPTION
**Pull request by LoserAntbear**
Closes #598 

## Changes:

- Add `AudioContext` And `AudioContextOperator` Singleton for handling multiple instances of `AudioContext`.
- Add `AudioInstance` with `WebAudioInstance` and `AudioTagInstance` implementations.
- Add `MediaEvent` which extends `GameEvent` with `NativeSoundEvent` implementation.
- `Sound` now extends `Resource` with all functionality. `Sound` now could be signed on:
  - 'volumechange' - on playing sound volume change;
  - 'pause' - on playback pause;
  - 'stop - on playback stop';
  - 'emptied' - on data cleanup;
  - 'resume' - on playback resume;
  - 'playbackstart' - on playback start;
  - 'playbackend' - on playback end;
- Move `WebAudio` to `Utils` accordingly to it's purpose.

Working on tests and `CHANGELOG` entries.

### CODE REVIEW

**Comment by LoserAntbear** [comment]
> expose native sudio dound events through excalibur 
That was rough night >.<

File: src/engine/Resources/Sound/AudioContext.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      }
+   }
+
+   private _getAudioCtxtId(ctxt: AudioContext): number {
+      const id = this._elements.indexOf(ctxt);
+
+      return (id > -1) ? id : null;
+   }
+
+   // private _getAudioCtxtByIndex(index: number): AudioContext {
```
[the remark this answers is not in the document]
Comment (LoserAntbear) [reply src/engine/Resources/Sound/AudioContext.ts]: Sorry, i sought it'd be useful to create this methods in advance, looks like they'll be needed some day. But okay, i'll remove them.

File: src/engine/Util/Util.ts [hunk 2]
```diff
@@ -1,7 +1,14 @@
 import { Vector } from '../Algebra';
 import { Random } from '../Math/Random';
 import { Side } from '../Collision/Side';
+import * as Sound from 'Sound';
 
+/**
+ * Export Sound
+ */
+export {
```
[the remark this answers is not in the document]
Comment (LoserAntbear) [reply src/engine/Util/Util.ts]: Yeah, i meant to export `canPlayFile` from `./Util/Sound`. I've changed comment to better understandability.

File: src/engine/Resources/Sound/AudioContext.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
+export class AudioContextFactory {
+   public static create() {
+      if ((<any>window).AudioContext || (<any>window).webkitAudioContext) {
```
[the remark this answers is not in the document]
Comment (eonarheim) [reply src/engine/Resources/Sound/AudioContext.ts]: I'm fairly certain we want only 1, we tried using multiple `AudioContext`s in the past. At the time browsers did not handle that well. 

A single `AudioContext` can play multiple concurrent audio streams, so there is no reason to support multiple.

File: src/engine/Events/GameEvent.ts [hunk 4]
```diff
@@ -1,85 +1,13 @@
...
-export type press = 'press';
-export type release = 'release';
-export type hold = 'hold';
-
-export type pointerdragstart = 'pointerdragstart';
-export type pointerdragend = 'pointerdragend';
-export type pointerdragenter = 'pointerdragenter';
-export type pointerdragleave = 'pointerdragleave';
-export type pointerdragmove = 'pointerdragmove';
+import { Scene } from '../Scene';
```
Comment (eonarheim) [inline src/engine/Events/GameEvent.ts]: Let's not refactor `Events.ts` at this time. It causes a lot of changes in files unrelated to audio.

File: src/engine/Resources/Sound/AudioContext.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (eonarheim) [reply src/engine/Resources/Sound/AudioContext.ts]: What would a `OfflineAudioContext` do?

File: src/engine/UIActor.ts [hunk 5]
```diff
@@ -21,7 +21,7 @@ export class UIActor extends Actor {
     * @param height  The starting height of the actor
     */
    constructor(xOrConfig?: number | IActorArgs, y?: number, width?: number, height?: number) {
-      if (typeof(xOrConfig !== 'object')) {
+      if (xOrConfig instanceof Object) {
```
Comment (eonarheim) [inline src/engine/UIActor.ts]: Let's revert changes to `UIActor`, this is unrelated to Audio

File: src/engine/Resources/Sound/AudioContext.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (LoserAntbear) [reply src/engine/Resources/Sound/AudioContext.ts]: Typically  `OfflineAudioContext` is used to pre-process audio data and than pass it to `AudioBuffer`. One of use-cases is described here:
When you want to create real-time canvas visualization (https://stackoverflow.com/questions/8074152/is-there-a-way-to-use-the-web-audio-api-to-sample-audio-faster-than-real-time).

and about multiple `AudioContexts`, maybe, we could create a pending issue, waiting for better support in browsers? Because `AudioContext` has useful methods like `suspend`, `resume`, etc. it could be useful while creating sound collections. Particularly, we could use multiple `AudioContext`s for issue #590 . Dividing sound in groups like "music", "ambient sounds", "sfx" etc. Handling each group in bounds of separate `AudioContext` would be pretty neat, i think.

**Comment by LoserAntbear** [comment]
So. Force-updated this one.
- Undid all changes related to moving of `Events.ts` to new namespace.
- Removed `AudioContextOperator` in favor of retuniing only single instance of `AudioContext` by `AudioContextFactory`
- And reverted back all Unrelated changes.

**Comment by LoserAntbear** [comment]
Fixed `CHANGELOG`  respectively.

**Comment by eonarheim** [comment]
@LoserAntbear Thanks again!

**Comment by LoserAntbear** [comment]
@eonarheim trying my best. Thank you, guys, too, btw, for letting me help you!
